### PR TITLE
PT-3435: Auditability of permissions and data sharing

### PR DESCRIPTION
**Pull request by veronikaslc**

### CODE REVIEW

**Comment by sdumitriu** [comment]
Jenkins, test this please.

**Comment by sdumitriu** [comment]
@veronikaslc Build is failing, can you fix it?

**Comment by veronikaslc** [comment]
fixed

**Comment by veronikaslc** [comment]
rebased @sashaandjic please, test

**Comment by sashaandjic** [comment]
1. The problem with public and open is now fixed. However, I'm not  sure if we should be showing the grid with users for them.
screen shot 2017-12-07 at 8 54 16 am
2. I'm still not sure about language in "The following X registered users...", as we also display groups in there. Once filters are applied the language isn't right either. I would go with something more generic. We should ask Brittney about this.
3. The Why column seems confusing and unnecessary. Can we simply add an "i" button next to the user/group name listed in the grid.
screen shot 2017-12-07 at 8 54 45 am

**Comment by marta-** [comment]
I agree with @sasha, I think the label text "why" is causing the confusion. It could be changed to "details" or removed completely. Also the icon (I) would work better than (?).

**Comment by marta-** [comment]
How about "the record can be accessed by:"?

**Comment by sashaandjic** [comment]
@marta +1

**Comment by veronikaslc** [comment]
@sashaandjic codereview comments addressed

**Comment by sashaandjic** [comment]
Works well on patients records.
Doesn't work on family records.
pt-3435

**Comment by veronikaslc** [comment]
@sashaandjic added families to the service

**Comment by sashaandjic** [comment]
Either remove the whole text or r

emove "patients" from "All administrator users have manage access to this patient record."

**Comment by sdumitriu** [comment]
@veronikaslc Merge conflicts, please resolve.

**Comment by veronikaslc** [comment]
@sdumitriu conflict resolved, rebased PR on master

**Comment by sashaandjic** [comment]
Is fixing "All administrator users have manage access to this patient record." also resolved?

**Comment by veronikaslc** [comment]
yes

**Comment by sashaandjic** [comment]
3435-family

**Comment by sashaandjic** [comment]
perfect. sorry about the previous comment. had to clear cache...

**Comment by sashaandjic** [comment]
I see the same problem and also noticed that pedigree isn't visible to John too. Is it possible that this problem isn;t related to this PR?

**Comment by sashaandjic** [comment]
For open and public records we show the grid, but hide filters. I would either show or hide both.
screen shot 2017-12-08 at 11 45 19 am

**Comment by veronikaslc** [comment]
@sdumitriu Should your comment be addressed now or later? "This will need to be refactored to be more generic once families are entity groups."

**Comment by sashaandjic** [comment]
@veronikaslc the problem that @sdumitriu reported still isn't addressed. I'll send you the dataset we use for testing on Slack. FAM0000004 has 4 patients. All patients have MaryGreen as owner and JohnDoe as collaborator. Even though FAM0000004 does not have JohnDoe as collaborator, JohnDoe can still view the family (he has view access to every family member) and he should be listed in the grid with view rights.

**Comment by veronikaslc** [comment]
@sashaandjic family members' owners added now, ready to final test

**Comment by sashaandjic** [comment]
Please remove "patient" from the info message.
screen shot 2017-12-12 at 11 23 42 am

**Comment by sashaandjic** [comment]
There is a problem in the following scenario:
- Mary is the owner of the family; Peter is collaborator with view only rights 
- Peter has edit rights to one of the patients
- Peter should be shown with edit rights even though on the family record he's assigned with view right

**Comment by sashaandjic** [comment]
build fails (without -Pquick); FAM0000004 shows
Mary as the only user with access rights (John should be there too)

File: components/entity-access-rules/rest/src/main/java/org/phenotips/data/permissions/rest/internal/DefaultDomainObjectFactory.java [hunk 1]
```diff
@@ -143,18 +191,264 @@ public CollaboratorsRepresentation createCollaboratorsRepresentation(PrimaryEnti
...
+                    checkAccessRight(addedPrincipals.get(ownerName).getLeft(), addedPrincipals.get(ownerName)
+                        .getRight(),
+                        this.editAccess, isFamily);
+                }
+            } else {
+                // - if this principal was never added and is Admin,
+                // then we save it save with access "manage" and role "admin".
+                DocumentReference ownerDocument = this.resolver.resolve(owner.getUser());
+                boolean isAdmin =
+                    this.rights.hasAccess(Right.ADMIN, ownerDocument, PermissionsConfiguration.PREFERENCES_DOCUMENT);
```
Comment (veronikaslc) [inline components/entity-access-rules/rest/src/main/java/org/phenotips/data/permissions/rest/internal/DefaultDomainObjectFactory.java]: @sdumitriu By some reason that throws null pointer exception. Did I misuse `hasAccess()` somehow?

File: components/entity-access-rules/rest/src/main/java/org/phenotips/data/permissions/rest/internal/DefaultDomainObjectFactory.java [hunk 2]
```diff
@@ -182,86 +204,239 @@ public PrincipalsRepresentation createPrincipalsRepresentation(PrimaryEntity ent
...
```
Comment (veronikaslc) [inline components/entity-access-rules/rest/src/main/java/org/phenotips/data/permissions/rest/internal/DefaultDomainObjectFactory.java]: @sdumitriu  By some reason that throws null pointer exception. Did I misuse hasAccess() somehow?

**Comment by sashaandjic** [comment]
tested. everything looks good except for data entered by guest users. 
for the quest uses's record we either don't need this functionality at all or the language needs to be slightly adjusted to:
1. "This record is owned by nobody" + (if collaborators exist "and can be accesses by")
2. "This record is public".
3. "This record is open."
4. "This record is owned by nobody and is hidden."
instead of:
1. 
private
2. 
public
3. 
open
4. 
hidden

**Comment by sashaandjic** [comment]
Registered users have no access to guest users records.
Change "This record is created by guest user and can be seen by anyone." to "This record is created by **a** guest user."
Everything else is good!

**Comment by veronikaslc** [comment]
@sashaandjic Wording changed

File: components/patient-access-rules/api/src/main/resources/ApplicationResources.properties [hunk 3]
```diff
@@ -63,7 +63,7 @@ phenotips.accessRightsManagement.modifyUpdateFailure=Failed to update.
 phenotips.accessRightsManagement.modifyRemoveCollaborator=Remove this collaborator
 
 phenotips.accessRightsManagement.adminInfo=All administrator users have manage access to this record.
-phenotips.accessRightsManagement.guestRecordInfo=This record is created by guest user and can be seen by anyone.
+phenotips.accessRightsManagement.guestRecordInfo=This record is created by a guest user.
```
Comment (sdumitriu) [inline components/patient-access-rules/api/src/main/resources/ApplicationResources.properties]: It should probably stay as it was, the fact that not everybody can see the record is a bug which I'm fixing right now.

**Comment by veronikaslc** [comment]
Wording change removed

**Comment by sashaandjic** [comment]
All good! The previous complaint was resolved in a separate PR.

File: components/patient-access-rules/api/src/main/resources/ApplicationResources.properties [hunk 4]
```diff
@@ -58,6 +58,30 @@ phenotips.accessRightsManagement.modifyNoServerResponse=Server not responding
...
+phenotips.accessRightsManagement.openRecordInfo=This record is open and can be modified by all registered users and groups.
+phenotips.accessRightsManagement.otherRecordInfo=This record is _VISIBILITY_ and can be accessed by:
+
+phenotips.accessRightsManagement.viewAccess=view
+phenotips.accessRightsManagement.editAccess=edit
+phenotips.accessRightsManagement.manageAccess=manage
+phenotips.accessRightsManagement.ownerRole=owner
+phenotips.accessRightsManagement.adminRole=administrator
+phenotips.accessRightsManagement.famMemOwnerRole=an owner of a family member
+phenotips.accessRightsManagement.famMemCollabRole=a collaborator of a family member
```
Comment (sdumitriu) [inline components/patient-access-rules/api/src/main/resources/ApplicationResources.properties]: collaborator on

File: components/patient-access-rules/api/src/main/resources/ApplicationResources.properties [hunk 5]
```diff
@@ -58,6 +58,30 @@ phenotips.accessRightsManagement.modifyNoServerResponse=Server not responding
 phenotips.accessRightsManagement.modifyUpdateFailure=Failed to update.
 phenotips.accessRightsManagement.modifyRemoveCollaborator=Remove this collaborator
 
+phenotips.accessRightsManagement.adminInfo=All administrator users have manage access to this record.
+phenotips.accessRightsManagement.guestRecordInfo=This record is created by guest user and can be seen by anyone.
```
Comment (sdumitriu) [inline components/patient-access-rules/api/src/main/resources/ApplicationResources.properties]: created by an unauthenticated user...
Also, can't it be modified by anyone?

**Pull request merged**